### PR TITLE
Meme page

### PR DESCRIPTION
**Pull request by erinlimbo**
- incorporate BlobStore
- initiated Meme page
- Created form that sends POST action to BlobStore by fetching

### CODE REVIEW

File: portfolio/src/main/webapp/meme.html [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      <script src="script.js"></script>
+        <div id="header">
+            <div class="headBox"> <a href="index.html"> Home </a> </div>
+            <div class="headBox"> <a href="team.html"> My Team </a> </div>
+            <div class="headBox"> <a href="meme.html"> Memes </a> </div>
+        </div>
+        <div id="content">
+            <h1> Welcome to the Meme Page! </h1>
+            <h3> (I love memes) </h3>
+
```
Comment (laskdaf) [inline portfolio/src/main/webapp/meme.html:19]: are you missing a `</div>` here?

File: portfolio/src/main/webapp/team.html [hunk 2]
```diff
@@ -11,6 +11,8 @@
         <div id="header">
             <div class="headBox"> <a href="index.html"> Home </a> </div>
             <div class="headBox"> <a href="team.html"> My Team </a> </div>
+            <div class="headBox"> <a href="meme.html"> Memes </a> </div>
+
```
Comment (laskdaf) [inline portfolio/src/main/webapp/team.html]: empty space

File: portfolio/src/main/webapp/meme.html (hunk 1 above)
Comment (erinlimbo) [reply portfolio/src/main/webapp/meme.html:19]: yessir

**Review by chaconnewu — APPROVED** [review]

**Comment by erinlimbo** [comment]
Made a lot of changes, can you unapprove the pull request in anyway?

File: portfolio/src/main/java/com/google/sps/data/Meme.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+   * @param {!String} desc The description of the meme
+   * @param {!String} timeStamp The time at the instant of initiation.
+   */
+  public Meme(long id, String author, String url, String desc, String timeStamp) {
+    this.id = id;
+    this.author = author;
+    this.url = url;
+    this.desc = desc;
+    this.timeStamp = timeStamp;
+}
```
Comment (laskdaf) [inline portfolio/src/main/java/com/google/sps/data/Meme.java]:
```suggestion
  }
```

File: portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  private int maxMemes = 40;
+
+  /** Read the data from the datastore and write it into /meme-handler as json. */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Meme").addSort("timeStamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    // List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxMemes));
```
Comment (laskdaf) [inline portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java]:
```suggestion
```

File: portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String description = request.getParameter("message");
+
+    String imageUrl = getUploadedFileUrl(request, "file");
+
+    Entity memeEntity = new Entity("Meme");
+    memeEntity.setProperty("author", "anonymous");
+    memeEntity.setProperty("url", imageUrl);
+    memeEntity.setProperty("desc", description);
+    memeEntity.setProperty("timeStamp", "11/22/11");
```
Comment (laskdaf) [inline portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java:84]: looks like this shouldn't be hardcoded

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -91,3 +91,39 @@ async function deleteComments() {
     });
     getComments();
 }
+
+/** Initiates the meme page */
+async function loadMemePage() {
+    let form = await fetchBlobstoreUrlAndShowForm();
+    let memes = await loadMemes();
```
Comment (laskdaf) [inline portfolio/src/main/webapp/script.js]: the functions don't look like they return anything, just do `await function()`

File: portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
```
Comment (laskdaf) [inline portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java]: should this be a todo?

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -91,3 +91,39 @@ async function deleteComments() {
...
+}
+
+/** Loads the memes from the datastore. */
+async function loadMemes() {
+    const memeContainer = document.getElementById("memes");
+    let response = await fetch('/meme-handler');
+    let data = await response.json();
```
Comment (laskdaf) [inline portfolio/src/main/webapp/script.js]: `const`?

**Review by laskdaf — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java (hunk 7 above)
Comment (erinlimbo) [reply portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java]: sorry i forgot to remove all of those before i pushed, ignore for now

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (erinlimbo) [reply portfolio/src/main/webapp/script.js]: Ah I ddint know I could do that

File: portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java (hunk 5 above)
Comment (erinlimbo) [reply portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java:84]: It shouldnt, gonna save incorporating users, and java.util.date in another pull request

Comment (laskdaf) [reply portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java:84]: can you not set it? or at least add a todo comment?

Comment (erinlimbo) [reply portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java:84]: gotcha

**Review by laskdaf — APPROVED** [review]

**Pull request merged**